### ISSUE TITLE
cvescan fails to report on (at least a set of kernel) security updates

### ISSUE THREAD

**Issue by kikoreis** [canonical/sec-cvescan#3]
I have an AWS instance (with no livepatch) which has the following updates available:
```
linux-aws/bionic-updates,bionic-security 4.15.0.1063.64 amd64 [upgradable from: 4.15.0.1060.61]
linux-headers-aws/bionic-updates,bionic-security 4.15.0.1063.64 amd64 [upgradable from: 4.15.0.1060.61]
linux-image-aws/bionic-updates,bionic-security 4.15.0.1063.64 amd64 [upgradable from: 4.15.0.1060.61]
```
However, cvescan doesn't report any CVEs applying:
```
ubuntu@ip-172-31-29-146:~$ cvescan -vp medium | tail
CVE-2016-1585
CVE-2015-8553
CVE-2014-4715
CVE-2013-7445
Running xsltproc to generate CVE list - fixable and filtered by priority
0 CVEs found with priority of medium or higher that can be fixed with package updates:

Full HTML report available in /home/ubuntu/snap/cvescan/common/report.htm
Normal non-verbose output will appear below
```

Thanks!

**Comment by kikoreis** [comment]
Note that the list above only shows the upgradable metapackages; the actual binaries aren't listed because due to how kernel package naming works (embedding versions) they are actually new packages, not upgrades. I wonder if that's related to this cvescan issue.

**Comment by kikoreis** [comment]
The hint as per the above was that motd correctly reports:

ubuntu@ip-172-31-29-146:~$ grep security /var/run/motd.dynamic 
7 updates are security updates.

The reason it knows is because /usr/lib/update-notifier/apt-check loops through everything which is marked either install or upgrade, i.e.:
```
            if not (depcache.marked_install(pkg) or
                    depcache.marked_upgrade(pkg)):
                continue
```
Whereas a simple apt list --upgradable won't list packages marked for installation.

**Comment by markmorlino** [comment]
I think this might be a problem with the OVAL generation rather than with cvescan.

ubuntu@ip-172-31-29-146:~$ grep 4.15.0.1063.64 /home/ubuntu/snap/cvescan/common/com.ubuntu.bionic.cve.oval.xml

And it might be related to the known problem we have where the OVAL uses the source package version (even in cases where the inary packages generated from that source have different version strings)

**Comment by markmorlino** [comment]
The OVAL is saying that it is fixed in a different version 

ubuntu@ip-172-31-29-146:~$ grep linux-aws /home/ubuntu/snap/cvescan/common/com.ubuntu.bionic.cve.oval.xml| grep test_ref | grep 1063 | head -n 1
                    <criterion test_ref="oval:com.ubuntu.bionic:tst:2019152170000010" comment="linux-aws package in bionic was vulnerable but has been fixed (note: '4.15.0-1063.67')." />

**Comment by stevebeattie** [comment]
Yes. Kernel's are particularly difficult to track because the meta packages that get updates in generated from a separate source package (linux-meta-aws in this case) from the actual source package that provides the binary packages (source linux-aws), and thus contains the actual fix. It gets even **more** complicated if signed kernels are used, because they are from a third source package (linux-signed-DERIVATIVE, would be linux-signed-aws on amd64 here).

In this case, the version @markmorlino is seeing above is the actual version of the linux-aws source package; but that differs from the version of the meta source package.

The oval does define:
```xml
        <linux-def:dpkginfo_object id="oval:com.ubuntu.bionic:obj:201245420000010" version="1" comment="The 'linux-aws' package binary.">
            <linux-def:name>linux-image-4.15.0-1063-aws</linux-def:name>
        </linux-def:dpkginfo_object\>
```
but that test object is not the one that is getting applied for the specific issues address by the 1063 kernel.

**Comment by stevebeattie** [comment]
Oh right, this is the test referenced above:
```xml
        <linux-def:dpkginfo_test id="oval:com.ubuntu.bionic:tst:2019152170000010" version="1" check_existence="at_least_one_exists" check="at least one" comment="Does the 'linux-aws' package exist and is the version less than '4.15.0-1063.67'?">
            <linux-def:object object_ref="oval:com.ubuntu.bionic:obj:201245420000010"/>
        <linux-def:state state_ref="oval:com.ubuntu.bionic:ste:2019152170000010" />
        </linux-def:dpkginfo_test>
```
which checks the object ```oval:com.ubuntu.bionic:obj:201245420000010```:
```xml                                                                                                                                                       
        <linux-def:dpkginfo_object id="oval:com.ubuntu.bionic:obj:201245420000010" version="1" comment="The 'linux-aws' package binary.">                    
            <linux-def:name>linux-image-4.15.0-1063-aws</linux-def:name>                                                                                     
        </linux-def:dpkginfo_object\>                                                                                                                        
```
in this case, the ```linux-image-4.15.0-1063-aws``` kernel package against the version ```4.15.0-1063.67```:
```xml
        <linux-def:dpkginfo_state id="oval:com.ubuntu.bionic:ste:2019152170000010" version="1" comment="The package version is less than '4.15.0-1063.67'.">
            <linux-def:evr datatype="debian_evr_string" operation="less than">4.15.0-1063.67</linux-def:evr>
        </linux-def:dpkginfo_state>
```

**Comment by stevebeattie** [comment]
But yes, what I believe (not having looked at the oval generator code) needs to happen as well is for it to lookup in launchpad what the specific matching abi version is for the associated meta package, and include a correct version test for that.

**Comment by markmorlino** [comment]
I think that explains why those are not being reported by cvescan. And the other discrepancy comes from update notifier showing both security packages that will be updated and packages that will be newly installed on the system but already have security updates. The latter won't be reported by cvescan since they are not installed yet.
